simplex/storage.go

```diff
@@ -0,0 +1,269 @@
+	db KeyValueReaderWriter
+
+	// genesisBlock is the genesis block data. It is stored as the first block in the storage.
+	genesisBlock *Block
+
+	// deserializer is used to deserialize quorum certificates from bytes.
+	deserializer *QCDeserializer
+
+	// blockTracker is used to manage blocks that have been indexed.
+	blockTracker *blockTracker
+
+	vm block.ChainVM
+
+	log logging.Logger
+}
+
+// newStorage creates a new prefixed database to store
+// finalizations according to their sequence numbers.
+// The VM is assumed to be initialized before calling this function.
+func newStorage(ctx context.Context, config *Config, qcDeserializer *QCDeserializer, blockTracker *blockTracker) (*Storage, error) {
+	lastAccepted, err := config.VM.LastAccepted(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	lastAcceptedBlock, err := config.VM.GetBlock(ctx, lastAccepted)
+	if err != nil {
+		return nil, err
+	}
+
+	genesisBlock, err := getGenesisBlock(ctx, config, blockTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Storage{
+		db:           config.DB,
+		genesisBlock: genesisBlock,
+		vm:           config.VM,
+		deserializer: qcDeserializer,
+		blockTracker: blockTracker,
+		log:          config.Log,
+	}
+	s.height.Store(lastAcceptedBlock.Height() + 1)
+
+	return s, nil
+}
+
+func (s *Storage) Height() uint64 {
+	return s.height.Load()
+}
+
+// Retrieve returns the block and finalization at [seq].
+// If [seq] is not found, returns false.
+func (s *Storage) Retrieve(seq uint64) (simplex.VerifiedBlock, simplex.Finalization, bool) {
+	if seq == 0 {
+		return s.genesisBlock, simplex.Finalization{}, true
+	}
+
+	block, err := getBlockAtHeight(context.TODO(), s.vm, seq)
+	if err != nil {
+		if err == database.ErrNotFound {
+			s.log.Error("block not found for sequence", zap.Uint64("seq", seq), zap.Error(err))
+		}
+		return nil, simplex.Finalization{}, false
+	}
+
+	finalization, found, err := s.retrieveFinalization(seq)
+	if !found || err != nil {
+		return nil, simplex.Finalization{}, false
+	}
+
+	vb := &Block{vmBlock: block, metadata: finalization.Finalization.ProtocolMetadata, blockTracker: s.blockTracker}
+	bytes, err := vb.Bytes()
+	if err != nil {
+		s.log.Error("failed to serialize block", zap.Error(err))
+		return nil, simplex.Finalization{}, false
+	}
+	vb.digest = computeDigest(bytes)
+
+	return vb, finalization, true
+}
+
+// Index indexes the finalization in the storage.
+// It stores the finalization bytes at the current height and increments the height.
+func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finalization simplex.Finalization) error {
+	bh := block.BlockHeader()
+	if bh.Seq == 0 {
+		s.log.Warn("attempted to index genesis block, which should not be indexed")
+		return errGenesisIndexed
+	}
+
+	currentHeight := s.height.Load()
+	if currentHeight != bh.Seq {
+		return fmt.Errorf("%w: expected %d, got %d", errUnexpectedSeq, currentHeight, bh.Seq)
+	}
+
+	if s.blockTracker.lastIndexed != bh.Prev {
```

Comment:
During review I initially found it odd that we reference `lastIndexed` without the lock held, but I assume that the expectation is that only this thread modifies `lastIndexed`, so it is safe to read this without the lock held.

Reply:
I don't think we should add this locking. I think a comment would make more sense.

Reply:
The only not concurrent thing in Simplex is block building and block verification, but that needs to be synchronized in the avalanchego side since the VM is not thread safe.